File: Projects/Regdll.pas
unit RegDLL;

{
  Inno Setup
  Copyright (C) 1997-2008 Jordan Russell
  Portions by Martijn Laan
  For conditions of distribution and use, see LICENSE.TXT.

  Registers 32-bit/64-bit DLL-based OLE servers in a child process (either
  regsvr32.exe or our own _RegDLL.tmp, depending on the Windows version)

  $jrsoftware: issrc/Projects/RegDLL.pas,v 1.7 2008/10/11 20:30:45 jr Exp $
}

interface

uses
  Windows;

procedure RegisterServer(const AUnregister: Boolean; const AIs64Bit: Boolean;
  const Filename: String; const AFailCriticalErrors: Boolean);

implementation

uses
  SysUtils, Forms, PathFunc, CmnFunc, CmnFunc2, InstFunc, Msgs, MsgIDs,
  Logging, RedirFunc, Main;

function WaitForAndCloseProcessHandle(var AProcessHandle: THandle): DWORD;
var
  WaitResult: DWORD;
begin
  try
    repeat
      { Process any pending messages first because MsgWaitForMultipleObjects
        (called below) only returns when *new* messages arrive }
      Application.ProcessMessages;
      WaitResult := MsgWaitForMultipleObjects(1, AProcessHandle, False, INFINITE, QS_ALLINPUT);
    until WaitResult <> WAIT_OBJECT_0+1;
    if WaitResult = WAIT_FAILED then
      Win32ErrorMsg('MsgWaitForMultipleObjects');
    if not GetExitCodeProcess(AProcessHandle, Result) then
      Win32ErrorMsg('GetExitCodeProcess');
  finally
    CloseHandle(AProcessHandle);
  end;
end;

procedure RegisterServerUsingRegSvr32(const AUnregister: Boolean;
  const AIs64Bit: Boolean; const Filename: String);
var
  SysDir, CmdLine: String;
  StartupInfo: TStartupInfo;
  ProcessInfo: TProcessInformation;
  ExitCode: DWORD;
begin
  SysDir := GetSystemDir;
  CmdLine := '"' + AddBackslash(SysDir) + 'regsvr32.exe"';
  if AUnregister then
    CmdLine := CmdLine + ' /u';
  CmdLine := CmdLine + ' /s "' + Filename + '"';
  if AIs64Bit then
    Log('Spawning 64-bit RegSvr32: ' + CmdLine)
  else
    Log('Spawning 32-bit RegSvr32: ' + CmdLine);

  FillChar(StartupInfo, SizeOf(StartupInfo), 0);
  StartupInfo.cb := SizeOf(StartupInfo);
  if not CreateProcessRedir(AIs64Bit, nil, PChar(CmdLine), nil, nil, False,
     CREATE_DEFAULT_ERROR_MODE, nil, PChar(SysDir), StartupInfo,
     ProcessInfo) then
    Win32ErrorMsg('CreateProcess');
  CloseHandle(ProcessInfo.hThread);
  ExitCode := WaitForAndCloseProcessHandle(ProcessInfo.hProcess);
  if ExitCode <> 0 then
    raise Exception.Create(FmtSetupMessage1(msgErrorRegSvr32Failed,
      Format('0x%x', [ExitCode])));
end;

{$IFNDEF UNICODE}
procedure RegisterServerUsingRegDLL(const AUnregister: Boolean;
  const Filename: String; const AFailCriticalErrors: Boolean);
type
  { This record is prepared and passed to the child process via shared memory.
    The child process modifies the 'out' fields as appropriate. }
  PRegProcessData = ^TRegProcessData;
  TRegProcessData = record
    Version: DWORD;                     { [in]  = RegProcessDataVersion }
    Size: DWORD;                        { [in]  = SizeOf(TRegProcessData) }
    Result: DWORD;                      { [out] Result of registration (rr*) }
    ResultCode: DWORD;                  { [out] Extended error code }
    Unregister: BOOL;                   { [in]  False if registering, True if unreg. }
    FailCriticalErrors: BOOL;           { [in]  Controls error mode }
    Filename: array[0..4095] of AnsiChar;   { [in]  DLL filename }
    Directory: array[0..4095] of AnsiChar;  { [in]  Directory to change to }
  end;
const
  RegProcessDataVersion = 3;

  { TRegProcessData Result values }
  rrOleInitializeFailed = 1;
  rrLoadLibraryFailed = 2;
  rrGetProcAddressFailed = 3;
  rrDllFunctionCalled = 4;

  { Process exit codes, full DWORD }
  ecSuccess = $1C9B28DA;

  RegFunctionNames: array[Boolean] of PChar =
    ('DllRegisterServer', 'DllUnregisterServer');

  InheritableSecurity: TSecurityAttributes = (
    nLength: SizeOf(InheritableSecurity); lpSecurityDescriptor: nil;
    bInheritHandle: True);
var
  TrueValue: BOOL;
  MutexHandle, MapHandle: THandle;
  Data: PRegProcessData;
  ExeFilename, CmdLine: String;
  StartupInfo: TStartupInfo;
  ProcessInfo: TProcessInformation;
  WaitResult, ExitCode: DWORD;
begin
  Log('Spawning _RegDLL.tmp');
  MapHandle := 0;
  Data := nil;
  { This TrueValue nonsense is to work around a bug in CreateMutex:
    it doesn't interpret -1 (what D3+ passes for True) as TRUE, only 1. }
  Integer(TrueValue) := 1;
  MutexHandle := CreateMutex(@InheritableSecurity, TrueValue, nil);
  if MutexHandle = 0 then
    Win32ErrorMsg('CreateMutex');
  try
    MapHandle := CreateFileMapping(INVALID_HANDLE_VALUE, @InheritableSecurity,
      PAGE_READWRITE, 0, SizeOf(Data^), nil);
    if MapHandle = 0 then
      Win32ErrorMsg('CreateFileMapping');
    Data := MapViewOfFile(MapHandle, FILE_MAP_WRITE, 0, 0, SizeOf(Data^));
    if Data = nil then
      Win32ErrorMsg('MapViewOfFile');

    FillChar(Data^, SizeOf(Data^), 0);
    Data.Version := RegProcessDataVersion;
    Data.Size := SizeOf(Data^);
    Data.Unregister := AUnregister;
    Data.FailCriticalErrors := AFailCriticalErrors;
    StrPLCopy(Data.Filename, Filename, (SizeOf(Data.Filename) div SizeOf(Data.Filename[0])) - 1);
    { For consistency with regsvr32-based registration, we no longer change
      to the directory containing the DLL: }
    //StrPLCopy(Data.Directory, PathExtractDir(Filename), (SizeOf(Data.Directory) div SizeOf(Data.Directory[0])) - 1);
    { We are the initial owner of the mutex. Release the mutex after all
      changes have been made to the record. This should impose a "release"
      memory barrier on architectures that require it. }
    if not ReleaseMutex(MutexHandle) then
      Win32ErrorMsg('ReleaseMutex');

    ExeFilename := AddBackslash(TempInstallDir) + '_isetup\_RegDLL.tmp';
    CmdLine := Format('_RegDLL.tmp %u %u', [MapHandle, MutexHandle]);
    FillChar(StartupInfo, SizeOf(StartupInfo), 0);
    StartupInfo.cb := SizeOf(StartupInfo);
    if not CreateProcess(PChar(ExeFilename), PChar(CmdLine), nil, nil,
       True, CREATE_DEFAULT_ERROR_MODE, nil, PChar(GetSystemDir),
       StartupInfo, ProcessInfo) then
      Win32ErrorMsg('CreateProcess');
    CloseHandle(ProcessInfo.hThread);
    ExitCode := WaitForAndCloseProcessHandle(ProcessInfo.hProcess);
    if ExitCode <> ecSuccess then
      InternalErrorFmt('REGDLL failed with exit code 0x%x', [ExitCode]);
    { Acquire the mutex ourself to ensure that all writes made to the record
      by the child process are visible to us in a multiprocessor scenario.
      (If we successfully obtain the mutex, then the child process has called
      ReleaseMutex, which imposes a "release" memory barrier, flushing prior
      writes.) }
    WaitResult := WaitForSingleObject(MutexHandle, INFINITE);
    if WaitResult <> WAIT_OBJECT_0 then
      InternalErrorFmt('REGDLL mutex wait failed (%d, %d)', [WaitResult, GetLastError]);

    case Data.Result of
      rrOleInitializeFailed: begin
          RaiseOleError('OleInitialize', Data.ResultCode);
        end;
      rrLoadLibraryFailed: begin
          Win32ErrorMsgEx('LoadLibrary', Data.ResultCode);
        end;
      rrGetProcAddressFailed: begin
          Win32ErrorMsgEx('GetProcAddress', Data.ResultCode);
        end;
      rrDllFunctionCalled: begin
          if FAILED(Data.ResultCode) then
            RaiseOleError(RegFunctionNames[AUnregister], Data.ResultCode);
        end;
    else
      InternalErrorFmt('REGDLL returned unknown result code %d', [Data.Result]);
    end;
  finally
    if Assigned(Data) then
      UnmapViewOfFile(Data);
    if MapHandle <> 0 then
      CloseHandle(MapHandle);
    CloseHandle(MutexHandle);
  end;
end;
{$ENDIF}

procedure RegisterServer(const AUnregister: Boolean; const AIs64Bit: Boolean;
  const Filename: String; const AFailCriticalErrors: Boolean);
var
  WindowDisabler: TWindowDisabler;
begin
  if AIs64Bit and not IsWin64 then
    InternalError('Cannot register 64-bit DLLs on this version of Windows');

  { Disable windows so the user can't utilize our UI while the child process
    is running }
  WindowDisabler := TWindowDisabler.Create;
  try
    { On Windows Vista, to get the "WRP Mitigation" compatibility hack which
      a lot of DLLs a require, we must use regsvr32.exe to handle the
      (un)registration.
      On Windows 2000/XP/2003, use regsvr32.exe as well for behavioral &
      error message consistency.
      On earlier versions, keep using RegDLL. Windows 95 and NT 4.0 (no SP)
      do not ship with regsvr32.exe. Windows 98 and Me do, but I prefer to
      treat all 9x OSes the same to simplify testing.
      Unicode build: Always use regsvr32.exe since there is no Unicode
      version of RegDLL, and we don't have to worry about pre-2000 Windows
      versions anyway. }
    {$IFDEF UNICODE}
    RegisterServerUsingRegSvr32(AUnregister, AIs64Bit, Filename);
    {$ELSE}
    if (WindowsVersion >= Cardinal($05000000)) or AIs64Bit then
      RegisterServerUsingRegSvr32(AUnregister, AIs64Bit, Filename)
    else
      RegisterServerUsingRegDLL(AUnregister, Filename, AFailCriticalErrors);
    {$ENDIF}
  finally
    WindowDisabler.Free;
  end;
end;

end.

File: Projects/RegDLL/Regdll.c
/*
  Inno Setup
  Copyright (C) 1997-2012 Jordan Russell
  Portions by Martijn Laan
  For conditions of distribution and use, see LICENSE.TXT.

  RegDLL process, called by RegDLL unit

  Compiled on Visual Studio 2005
*/

#include <windows.h>
#include <commctrl.h>

typedef struct {
	DWORD Version;
	DWORD Size;
	DWORD Result;
	DWORD ResultCode;
	BOOL Unregister;
	BOOL FailCriticalErrors;
	CHAR Filename[4096];
	CHAR Directory[4096];
} REG_PROCESS_DATA;

#define REG_PROCESS_DATA_VERSION 3

// REG_PROCESS_DATA Result values
#define rrOleInitializeFailed 1
#define rrLoadLibraryFailed 2
#define rrGetProcAddressFailed 3
#define rrDllFunctionCalled 4

// Process exit codes, full DWORD
#define ecSuccess 0x1C9B28DA

// Process exit codes, high word
#define ecMapFailed 1
#define ecMutexWaitFailed 2
#define ecMutexWaitUnexpectedResult 3
#define ecWrongVersionOrSize 4

LPTSTR GetCommandTail(void)
{
	LPTSTR CmdLine = GetCommandLine();
	BOOL InQuote = FALSE;
	while (*CmdLine) {
		if (!InQuote && *CmdLine <= ' ') break;
		if (*CmdLine == '"') InQuote = !InQuote;
		CmdLine++;
	}
	while (*CmdLine && *CmdLine <= ' ') {
		CmdLine++;
	}
	return CmdLine;
}

LPTSTR ExtractHandleValue(LPTSTR CmdLine, HANDLE *OutHandle)
{
	UINT_PTR Result = 0;
	while (*CmdLine >= '0' && *CmdLine <= '9') {
		Result *= 10;
		Result += (*CmdLine - '0');
		CmdLine++;
	}
	while (*CmdLine && *CmdLine <= ' ') {
		CmdLine++;
	}
	*OutHandle = (HANDLE)Result;
	return CmdLine;
}

LONG Main(void)
{
	LPTSTR Args;
	HANDLE MapHandle, MutexHandle;
	REG_PROCESS_DATA *Data;
	DWORD WaitResult;
	HRESULT OleInitResult;
	HMODULE LibHandle;

	// Work around bug in Windows XP Gold & SP1: If the application manifest
	// specifies COMCTL32.DLL version 6.0 (to enable visual styles), we must
	// call InitCommonControls() to ensure that we actually link to
	// COMCTL32.DLL, otherwise calls to MessageBox() fail. (XP SP2 appears
	// to fix this.)
	InitCommonControls();

	// Extract the two handle values from the command line
	Args = GetCommandTail();
	Args = ExtractHandleValue(Args, &MapHandle);
	ExtractHandleValue(Args, &MutexHandle);

	// Map shared section
	Data = MapViewOfFile(MapHandle, FILE_MAP_WRITE, 0, 0, sizeof(*Data));
	if (Data == NULL) {
		return MAKELONG(GetLastError(), ecMapFailed);
	}

	// Acquire mutex
	WaitResult = WaitForSingleObject(MutexHandle, INFINITE);
	if (WaitResult == WAIT_FAILED) {
		return MAKELONG(GetLastError(), ecMutexWaitFailed);
	}
	if (WaitResult != WAIT_OBJECT_0) {
		return MAKELONG(WaitResult, ecMutexWaitUnexpectedResult);
	}

	// Check structure version number and size
	if (Data->Version != REG_PROCESS_DATA_VERSION ||
		Data->Size != sizeof(*Data)) {
		return MAKELONG(0, ecWrongVersionOrSize);
	}

	// Initialize OLE. Regsvr32 does this, and it's needed when registering
	// comcat.dll on stock Windows 98. Its DllRegisterServer export
	// forwards to ole32.dll; if ole32.dll isn't loaded, GetProcAddress
	// fails.
	OleInitResult = OleInitialize(NULL);
	if (SUCCEEDED(OleInitResult)) {
		// Set error mode
		SetErrorMode(Data->FailCriticalErrors ?
			SEM_NOOPENFILEERRORBOX | SEM_FAILCRITICALERRORS :
			SEM_NOOPENFILEERRORBOX);

		// Initialize current directory
		if (*Data->Directory) {
			SetCurrentDirectoryA(Data->Directory);
		}

		// Load & register the DLL
		LibHandle = LoadLibraryExA(Data->Filename, NULL, LOAD_WITH_ALTERED_SEARCH_PATH);
		if (LibHandle) {
			typedef HRESULT (STDAPICALLTYPE *DLLREGISTERSERVER)(void);
			DLLREGISTERSERVER RegisterServerProc;

			RegisterServerProc = (DLLREGISTERSERVER)GetProcAddress(LibHandle,
				Data->Unregister ? "DllUnregisterServer" : "DllRegisterServer");
			if (RegisterServerProc) {
				Data->ResultCode = (*RegisterServerProc)();
				Data->Result = rrDllFunctionCalled;
			} else {
				Data->ResultCode = GetLastError();
				Data->Result = rrGetProcAddressFailed;
			}

			FreeLibrary(LibHandle);
		} else {
			Data->ResultCode = GetLastError();
			Data->Result = rrLoadLibraryFailed;
		}

		OleUninitialize();
	} else {
		Data->ResultCode = OleInitResult;
		Data->Result = rrOleInitializeFailed;
	}

	// Release the mutex after all changes have been made to the record. This
	// should impose a "release" memory barrier on architectures that require
	// it.	
	ReleaseMutex(MutexHandle);

	return ecSuccess;
}

int WinMainCRTStartup(void)
{
	LONG Result = Main();

	// Good idea to call ExitProcess because it'll terminate any other
	// threads the system might've created. A simple "return" won't.
	ExitProcess(Result);

	return Result;
}